src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,151 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/SamplingLoader.hpp>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/pool.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <cast_core/helpers/pipeline_helpers.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/PipelineGenerator.hpp>
+
+namespace genny::actor {
+
+/** @private */
+using index_type = std::tuple<DocumentGenerator, std::optional<DocumentGenerator>, std::string>;
+using namespace bsoncxx;
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+/** @private */
+struct SamplingLoader::PhaseConfig {
+    PhaseConfig(PhaseContext& context, mongocxx::pool::entry& client, ActorId id)
+        : database{(*client)[context["Database"].to<std::string>()]},
+          collection{database[context["Collection"].to<std::string>()]},
+          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()},
+          // A sample size is optional. If unspecified, we'll make a sample size such that each
+          // sample document is inserted one time.
+          sampleSize{
+              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
+          pipelineSuffixGenerator{context["Pipeline"]
+                                      .maybe<PipelineGenerator>(context, id)
+                                      .value_or(PipelineGenerator{})} {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    // User can specify any pipeline here, but it's worth noting that some will not play nicely. For
+    // example, adding a $match will deflate the sample size.
+    PipelineGenerator pipelineSuffixGenerator;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(debug) << "Beginning to run SamplingLoader";
+            // Read the sample size documents into a vector.
+            mongocxx::pipeline samplePipeline;
+            samplePipeline.sample(config->sampleSize);
+            samplePipeline.project(make_document(kvp("_id", 0)));
+            const auto suffixStages =
+                pipeline_helpers::makePipeline(config->pipelineSuffixGenerator).view_array();
+            samplePipeline.append_stages(suffixStages);
+            auto cursor =
+                config->collection.aggregate(samplePipeline, mongocxx::options::aggregate{});
+            const auto sampleDocs =
```

Comment:
Depending on the sample size and the docs in the collection this might use quite a bit of memory. Is it a concern? For example, might it interfere with the SUT by creating memory pressure? A similar question for `batchOfDocs`. What is the "reasonable" range for the config options for this actor we expect? Should it be documented in the SamplingLoader.yml as recommended?

Reply:
I meant to respond to this but forgot! I think it's unlikely to be a problem in evergreen, since I'm pretty sure we're smart enough to run the client on a separate physical node than the server. Otherwise we'd be competing on compute as well for high thread counts. I still think it's worth a warning though, since you might not guess we'd store them all in memory at once.